OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs

```diff
@@ -150,5 +150,73 @@ protected static bool ReloadsAutomatically(IEnumerable<AmmoPool> ammoPools, Rear
 
 			return true;
 		}
+
+		// Retreat units from combat, or for supply only in idle
+		protected void Retreat(Squad squad, bool flee, bool rearm, bool repair)
+		{
+			var loc = new CPos(0, 0, 0);
+
+			// HACK: "alreadyRepair" is to solve AI repair orders performance,
+			// which is only allow one goes to repairpad at the same time to avoid queueing too many orders.
+			// if repairpad logic is better we can just drop it.
+			var alreadyRepair = false;
+
+			if (flee)
+				loc = RandomBuildingLocation(squad);
+
+			foreach (var a in squad.Units)
+			{
+				if (IsRearming(a))
+					continue;
+
+				var orderQueued = false;
+
+				// Try rearm units.
+				if (rearm)
+				{
+					var ammoPools = a.TraitsImplementing<AmmoPool>().ToArray();
+					if (!ReloadsAutomatically(ammoPools, a.TraitOrDefault<Rearmable>()) && !FullAmmo(ammoPools))
+					{
+						squad.Bot.QueueOrder(new Order("ReturnToBase", a, orderQueued));
+						orderQueued = true;
+					}
+				}
+
+				// Try repair units.
+				if (repair && !alreadyRepair)
+				{
+					Actor repairBuilding = null;
+					var orderId = "Repair";
+					var health = a.TraitOrDefault<IHealth>();
+
+					if (health != null && health.DamageState > DamageState.Undamaged)
+					{
+						var repairable = a.TraitOrDefault<Repairable>();
+						if (repairable != null)
+							repairBuilding = repairable.FindRepairBuilding(a);
+						else
+						{
+							var repairableNear = a.TraitOrDefault<RepairableNear>();
+							if (repairableNear != null)
+							{
+								orderId = "RepairNear";
+								repairBuilding = repairableNear.FindRepairBuilding(a);
+							}
+						}
+
+						if (repairBuilding != null)
```

Comment:
```suggestion
						if (repairBuilding != null) && Target.FromActor(repairBuilding).IsValidFor(a)
```

may fix https://github.com/OpenRA/OpenRA/pull/18801 due to

https://github.com/OpenRA/OpenRA/blob/75fe0e524fb8013664f38027bfffb9ad6f99799d/OpenRA.Mods.Common/Traits/Buildings/Reservable.cs#L39-L45

invalidating it every tick.

Reply:
If so there is also a crash threat in RepairOrderGenerator.cs

Reply:
Ummm..
Crash still happens here.

Reply:
···
OpenRA engine version {DEV_VERSION}
Tiberian Sun mod version {DEV_VERSION}
on map 28a09cab7f1ee45c4ea5b339c85ce41b3d5fb9c3 (Way to Uganda by Morton).
Date: 2020-11-03 15:27:41Z
···